MAL implementation in Modula-2.

zawaza-blog opened the pull request:
This implementation uses the GNU Modula2 compiler and it runs all sets and test suites. Dockerfile is availabe for systems without a gm2 compiler.

The implemention was harder than I expected. I worked with modula-2 twice in my early jobs. And I have fond memories of it. This work has also served as a reminder of those early days when everything was yet to be built, and we reinvented the wheel continously.

Summary
-------

* All steps implemented, Pass all tests.
* Do not use modula2 ALLOCATE, details below.
* Do not have a garbage collector.

GM2 Compiler
------------
Having a GM2 compiler was the key enabler of this implementation, furthermore it includes the PIM and ISO libraries.

I realized there was a modula-2 compiler two years ago, despite it have been available for more than ten years. Knowing the compiler was available, the only remaining challenge was finding enough time to dedicate to the project.

GM2 ALLOCATE/NEW
----------------
50% of the project was fighting with the ALLOCATE/NEW to reserve memory. It's slow. ALLOCATE / NEW also tracks every reserved object in a list. Some of the steps make heavy use of dynamic memory, and the ALLOCATE is really slow due to the objet tracking management. By step 5 i give up, and started using malloc. It was soo slow that the TCO test timeout despite the tail call optimization was workin was working..

Garbage collector
-----------------
I intend to implement a Garbage Collector (Mark-and-Sweep or Cheney-style copying), if I can allocate additional time.

Run Suite
---------
$ make "docker-build^modula2"
$ make DOCKERIZE=1 "test^modula2"

Thanks
------
Obviously this implementation wasn't posible without the work of Niklaus Wirth and Gaius Mulley.

Pull request requirements:

- [ ] Commits are well written and well organized.
- [ ] Commits for a specific implementation should be prefixed with
  the implementation name.
- [ ] Github Actions CI passes all checks (including self-host)

Additional requirements if you are adding a new implementation (see [FAQ](../docs/FAQ.md#add_implementation) for details):

- [ ] Follow incremental structure (no common eval code)
- [ ] Add `impls/<IMPL>/Dockerfile`
- [ ] Add `impls/<IMPL>/Makefile`
- [ ] Update `IMPLS.yml`
- [ ] Update `Makefile.impls`
- [ ] Update `README.md`

zawaza-blog committed:
MAL implementation in Modula-2.

This implementation uses the GNU Modula2 compiler and
it runs all sets and test suites. Dockerfile is availabe for systems without a gm2 compiler.

The implemention was harder than I expected. I worked with modula-2 twice in my early jobs.
And I have fond memories of it. This work has also served as a reminder of those early days
when everything was yet to be built, and we reinvented the wheel continously.

Summary
-------

* All steps implemented, Pass all tests.
* Do not use modula2 ALLOCATE, details below.
* Do not have a garbage collector.

GM2 Compiler
------------
Having a GM2 compiler was the key enabler of this implementation, furthermore it includes the PIM and ISO libraries.

I realized there was a modula-2 compiler two years ago, despite it have been available for more than ten years.
Knowing the compiler was available, the only remaining challenge was finding enough time to dedicate to the project.

GM2 ALLOCATE/NEW
----------------
50% of the project was fighting with the ALLOCATE/NEW to reserve memory. It's slow.
ALLOCATE / NEW also tracks every reserved object in a list. Some of the steps make heavy use of dynamic memory, and the
ALLOCATE is really slow due to the objet tracking management. By step 5 i give up, and started using malloc.
It was soo slow that the TCO test timeout despite the tail call optimization was workin was working..

Garbage collector
-----------------
I intend to implement a Garbage Collector (Mark-and-Sweep or Cheney-style copying), if I can allocate additional time.

Run Suite
---------
$ make "docker-build^modula2"
$ make DOCKERIZE=1 "test^modula2"

Thanks
------
Obviously this implementation wasn't posible without the work of Niklaus Wirth and Gaius Mulley.

zawaza-blog committed:
Alignment with Commit 2bbfaa5. Use /usr/bin/env bash shebangs instead of /bin/bash.

kanaka commented:
Hi @zawaza-blog. Sorry for the slow reply (work, travel, holidays, etc). Can you please add your implementation to `IMPLS.yml` so that the automated CI will activate?

zawaza-blog committed:
modula2: Added modula2 to the IMPLS.yml so that the automated CI will activate.

zawaza-blog committed:
modula2: fixes for self-hosted tests.

- step2. input/output error.
- stepA. EVAL checked HasError() but not HasException() in recursive calls, throw  didn't stop evaluation.

zawaza-blog commented:
Hi @kanaka I'm really grateful you have find time to look to the pull requests. 

The modula2 implementation has been added to IMPLS.yml, and the issues triggered in the self-hosted CI had been solved.

kanaka commented on impls/mal/run:
```diff
@@ -6,4 +6,4 @@ case ${MAL_IMPL} in
     MAL_FILE="../mal/stepA_mal.mal ${MAL_FILE}" ;;
 esac
-exec ./../${MAL_IMPL:-js}/run ${MAL_FILE} "${@}"
+exec ./../${MAL_IMPL:-modula2}/run ${MAL_FILE} "${@}"
```
I think you accidentally changed the mal run script to use modula2 by default. You shouldn't need to do this. If you run `make MAL_IMPL=modula2 test^mal` that should run the tests is self-host mode using the modula2 implementation as the host implementation. The "js" here is just the fallback if you don't specify MAL_IMPL.

zawaza-blog replied:
Ops. Sorry. default MAL_IMPL has been reverted to "js".

Thanks again for your time and dedication. 😃

kanaka reviewed (commented):
@zawaza-blog Looks good. One inline comment that needs to be fixed before I merge it.

zawaza-blog committed:
modula2: reverted default implementation to js in impls/mal/run.

kanaka commented:
Hi @zawaza-blog. I'm requesting a couple more changes after reviewing the code in more detail:
* Remove the debug at the end of step5
* Use the current approach in the guide for how to handle macroexpand. This implementation is currently using the prior more complicated approach that has a separate macroexpand function/phase. The current approach integrates macroexpansion into the EVAL function in a more simple way.

zawaza-blog commented:
Of course :smiley:, I will try to find some time to align steps 8 and 9 with the current approach of macroexpansion.

zawaza-blog committed:
modula2: align macroexpand with current description in step8.

- step8, step9 & stepA
  * Removed the early macro expansion in eval.
  * Changes to function application.
    - First evaluate only the first element. Shall be a function.
    - If function^.isMacro, apply it to unevaled arguments and update ast.
    - Else evaluate remaining ast and apply the function.

- Removed the env & cache debug at the end of step5

zawaza-blog commented:
Finally, I found spare time to implement the macroexpand current approach and cleanup logs. It was fun. 😄

kanaka commented:
@zawaza-blog Work and travel continues to be busy so my responses here are slow. I had a chance to review the latest changes. It looks like the unused macroexpand and macro test functions are still in the code (the new approach should result in substantial code size reduction to steps 8, 9, and A).

BTW, there are some things about the code that make me think substantial AI/LLM help was used to generate this. That's not a blocker to merging this (I think by the end of this year hand-crafted code will be the exception). However, it's an interesting data point to me, so could you describe what type/level of LLM help you used for this?

kanaka commented:
@zawaza-blog Still want to pursue this. I think there just a couple minor things to update and answer (see my last comment).

zawaza-blog commented:
Yeah,  I'm still interested. I have being quite busy latelly, but on hollidays i will find spare time.   

Regarding your question about IA use. By fall 2025 I do a big refactoring. Mostly changing default allocator to libc & malloc.

zawaza-blog committed:
modula2: clean unused "is macro" code.

zawaza-blog committed:
modula2: clean of unused code.
